core/templates/dev/head/player/InteractionService.js

```diff
@@ -0,0 +1,94 @@
+ *
+ * @author wxyxinyu@gmail.com (Xinyu Wu)
+ */
+
+oppia.constant('DEFAULT_ANSWER_GROUP_CLASSIFICATION_THRESHOLD', 0.3);
+
+oppia.factory('interactionService', [
+    'warningsData', 'DEFAULT_ANSWER_GROUP_CLASSIFICATION_THRESHOLD', function(
+    warningsData, DEFAULT_ANSWER_GROUP_CLASSIFICATION_THRESHOLD) {
+  var stringifyClassifiedRule = function(rule) {
+    var paramList = [];
+    for (var key in rule.inputs) {
+      paramList.push(rule.inputs[key]);
+    }
+    var stringifiedRule = rule.rule_type;
+    stringifiedRule += '(' + paramList.join(',') + ')';
+    return stringifiedRule;
+  };
+  return {
+    'classifyAnswer': function(
+      answer, answerClassificationData, RULES) {
+      var answerGroups = answerClassificationData.answerGroups;
+      var defaultOutcome = answerClassificationData.defaultOutcome;
+
+      // Find the first group that satisfactorily matches the given answer. This
+      // is done by ORing (maximizing) all truth values of all rules over all
+      // answer groups. The group with the highest truth value is considered the
+      // best match.
+      var bestMatchedAnswerGroup;
+      var bestMatchedAnswerGroupIndex = answerGroups.length;
+      var bestMatchedRuleSpec;
+      var bestMatchedTruthValue = 0.0;
+      for (var i = 0; i < answerGroups.length; i++) {
+        var oredTruthValue = 0.0;
+        var bestRuleSpec;
+        for (var j = 0; j < answerGroups[i].rule_specs.length; j++) {
+          var ruleSpec = answerGroups[i].rule_specs[j];
+          var evaluatedTruthValue = RULES[ruleSpec.rule_type](
+            answer, ruleSpec.inputs);
+          if (evaluatedTruthValue > oredTruthValue) {
+            oredTruthValue = evaluatedTruthValue;
+            bestRuleSpec = ruleSpec;
+          }
+        }
+        if (oredTruthValue > bestMatchedTruthValue) {
+          bestMatchedTruthValue = oredTruthValue;
+          bestMatchedRuleSpec = bestRuleSpec;
+          bestMatchedAnswerGroup = answerGroups[i];
+          bestMatchedAnswerGroupIndex = i;
+        }
+      }
+
+      // The best matched group must match above a certain threshold. If no
+      // group meets this requirement, then the default 'group' automatically
+      // matches resulting in the outcome of the answer being the default
+      // outcome of the state.
+      var DEFAULT_RULESPEC_STRING = 'Default';
+      if (bestMatchedTruthValue >=
+          DEFAULT_ANSWER_GROUP_CLASSIFICATION_THRESHOLD) {
+        return {
+          'outcome': bestMatchedAnswerGroup.outcome,
+          'rule_spec_string': stringifyClassifiedRule(bestMatchedRuleSpec),
+          'answer_group_index': bestMatchedAnswerGroupIndex,
+          'classification_certainty': bestMatchedTruthValue
+        };
+      } else if (defaultOutcome !== undefined) {
+        return {
+          'outcome': defaultOutcome,
+          'rule_spec_string': DEFAULT_RULESPEC_STRING,
+          'answer_group_index': answerGroups.length,
```

Comment:
Clarified on classification_certainty above (it's not needed). answer_group_index will especially be needed after you get rid of rule_spec_string since the backend will need it to figure out whether to use the default rule spec string or to lookup the actual RuleSpec from which to generate a rulespec string.

The use of answer_group_index with training unresolved answers is outside the scope of this PR, but it may come up in the future as Stephen's work is merged in, the answer model is rebuilt, and we are able to turn on the training interface.